## 1. Importing libraries

In [2]:
import os
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

## 2.Reading the Data


In [3]:
PROJECT_DIR=r"D:\DS_ML\Flight-AWS_sagemaker_project"
DATA_DIR="Data"


In [4]:
def get_data(name):
    file_name=f"{name}.csv"
    file_path=os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    return pd.read_csv(file_path)

In [5]:
flights=get_data("flight_price")
flights

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [6]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- The dataset contains 10683 rows and 11 features
- Colums "ROUTE" AND "Total stops" have missing value
- The datatype of few feature are not appropriate

## 3. Preliminary analysis

### Observations

- The type of "Date_of_Journey " ,"Dep_Time" and "Arrival_Time" should be changed to datetime
- The type of "duration" and "totalstops" is mixed.It should be numeric 
- There are 220 duplicates that has to be removed


### 3.1Check Datatypes 

In [7]:
flights["Date_of_Journey"].iloc[6]


'12/03/2019'

In [8]:
flights["Duration"].iloc[5]

'2h 25m'

### 3.2 Check for Duplictaes


In [9]:
flights.duplicated().sum()

220

In [10]:
(
flights
.loc[flights.duplicated(keep=False)]
.sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

## 4. Detailed Analysis

In [11]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [12]:
(
flights
.Airline
.str.replace(" Premium economy","")
.str.replace(" Business","")
.str.title()
.unique()
)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Trujet'],
      dtype=object)

- Some of the entries are class not airlines

In [13]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [14]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [15]:
flights[flights["Destination"]=="Delhi"]

Airline Date_of_Journey    Source Destination      Route Dep_Time  \
22          IndiGo       3/04/2019  Banglore       Delhi  BLR → DEL    04:00   
23          IndiGo       1/05/2019  Banglore       Delhi  BLR → DEL    18:55   
28         Vistara      18/06/2019  Banglore       Delhi  BLR → DEL    09:45   
32          IndiGo       6/04/2019  Banglore       Delhi  BLR → DEL    04:00   
43        Air Asia       6/05/2019  Banglore       Delhi  BLR → DEL    11:10   
...            ...             ...       ...         ...        ...      ...   
10658  Jet Airways      18/06/2019  Banglore       Delhi  BLR → DEL    11:10   
10659      Vistara       1/05/2019  Banglore       Delhi  BLR → DEL    21:00   
10664       IndiGo       9/04/2019  Banglore       Delhi  BLR → DEL    18:55   
10677     SpiceJet      21/05/2019  Banglore       Delhi  BLR → DEL    05:55   
10680  Jet Airways      27/04/2019  Banglore       Delhi  BLR → DEL    08:20   

      Arrival_Time Duration Total_Stops               Additional_Info  Price  
22           06:50   2h 50m    non-stop                       No info   3943  
23           21:50   2h 55m    non-stop                       No info   4823  
28           12:35   2h 50m    non-stop                       No info   4668  
32           06:50   2h 50m    non-stop                       No info   4423  
43           13:55   2h 45m    non-stop                       No info   3383  
...            ...      ...         ...                           ...    ...  
10658        14:05   2h 55m    non-stop   In-flight meal not included   5769  
10659        23:50   2h 50m    non-stop                       No info   4668  
10664        21:50   2h 55m    non-stop                       No info   4823  
10677        08:35   2h 40m    non-stop  No check-in baggage included   3257  
10680        11:20       3h    non-stop                       No info   7229  

[1265 rows x 11 columns]

In [16]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [17]:
(
flights
.Dep_Time
.loc[lambda ser:ser.str.contains('[^0-9:]')]
)

Series([], Name: Dep_Time, dtype: object)

In [18]:
(
flights
.Arrival_Time
.loc[lambda ser:ser.str.contains('[^0-9:]')]
.str.split(" ",n=1)
.str.get(1)
.unique()    
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

In [19]:
flights.Duration.unique()

array(['2h 50m', '7h 25m', '19h', '5h 25m', '4h 45m', '2h 25m', '15h 30m',
       '21h 5m', '25h 30m', '7h 50m', '13h 15m', '2h 35m', '2h 15m',
       '12h 10m', '26h 35m', '4h 30m', '22h 35m', '23h', '20h 35m',
       '5h 10m', '15h 20m', '2h 55m', '13h 20m', '15h 10m', '5h 45m',
       '5h 55m', '13h 25m', '22h', '5h 30m', '10h 25m', '5h 15m',
       '2h 30m', '6h 15m', '11h 55m', '11h 5m', '8h 30m', '22h 5m',
       '2h 45m', '12h', '16h 5m', '19h 55m', '3h 15m', '25h 20m', '3h',
       '16h 15m', '15h 5m', '6h 30m', '25h 5m', '12h 25m', '27h 20m',
       '10h 15m', '10h 30m', '1h 30m', '1h 25m', '26h 30m', '7h 20m',
       '13h 30m', '5h', '19h 5m', '14h 50m', '2h 40m', '22h 10m',
       '9h 35m', '10h', '21h 20m', '18h 45m', '12h 20m', '18h', '9h 15m',
       '17h 30m', '16h 35m', '12h 15m', '7h 30m', '24h', '8h 55m',
       '7h 10m', '14h 30m', '30h 20m', '15h', '12h 45m', '10h 10m',
       '15h 25m', '14h 5m', '20h 15m', '23h 10m', '18h 10m', '16h',
       '2h 20m', '8h', '16h 5

In [20]:
(
flights
.Duration
.loc[lambda ser: ~ser.str.contains("h")]
)

6474    5m
Name: Duration, dtype: object

In [21]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

- this observation 6474 is completely wrong

In [22]:
(
flights
.Duration
.drop(index=6474)
.str.split(" ",expand=True)
.set_axis(["hour","minute"], axis=1)
.assign(
hour=lambda df_:(
    df_
    .hour
    .str.replace("h","")
    .astype(int)
    .mul(60)
),
    minute=lambda df_:(
    df_
    .minute
    .str.replace("m","")
    .fillna("0")
    .astype(int)
    )
)
.sum(axis=1)
)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

In [23]:
(
flights
.Duration
.drop(index=6474,index)
.str.split(" ",expand=True)
.set_axis(["hour","minute"], axis=1)
.assign(
hour=lambda df_:(
    df_
    .hour
    .str.replace("h","")
    .astype(int)
    .mul(60)
),
    minute=lambda df_:(
    df_
    .minute
    .str.replace("m","")
    .fillna("0")
    .astype(int)
    )
)
.sum(axis=1)
.rename("duration minutes")
.to_frame()
.join(flights.Duration.drop(index=6474))
)

duration minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

In [24]:
def convert_to_minutes(ser):
    return (
    ser
    .str.split(" ",expand=True)
    .set_axis(["hour","minute"], axis=1)
    .assign(
    hour=lambda df_:(
        df_
        .hour
        .str.replace("h","")
        .astype(int)
        .mul(60)
    ),
        minute=lambda df_:(
        df_
        .minute
        .str.replace("m","")
        .fillna("0")
        .astype(int)
        )
    )
    .sum(axis=1))
    

In [25]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [28]:
(
    flights
    .Total_Stops
    .replace("non-stop","0")
    .str.replace(" stops?","",regex=True)
    .pipe(lambda ser:pd.to_numeric(ser))
    
)

array([ 0.,  2.,  1.,  3., nan,  4.])

In [30]:
flights.Total_Stops.isna().sum()

1

In [29]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [32]:
flights[flights['Total_Stops'].isna()==True]

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

## 5.Cleaning Operation

In [33]:
def clean_data(df):
    return(
    df
     .drop(index=[6474,9039])
     .drop_duplicates() 
     .assign(**{
         col: df[col].str.strip()
         for col in df.select_dtypes(include="O").columns
     }
    )    
    .rename(columns=str.lower)
    .assign(
        airline=lambda df_:(
            df_
            .airline
            .str.replace(" Premium economy","")
            .str.replace(" Business","")
            .str.title()
            ),
        date_of_journey=lambda df_:pd.to_datetime(df_.date_of_journey,dayfirst=True),
        dep_time=lambda df_:pd.to_datetime(df_.dep_time).dt.time,
        arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,
        duration=lambda df_:df_.duration.pipe(convert_to_minutes),
        total_stops=lambda df_:(
            df_
            .total_stops
            .replace("non-stop","0")
            .str.replace(" stops?","",regex=True)
            .pipe(lambda ser:pd.to_numeric(ser))
            ),
        additional_info=lambda df_:df_.additional_info.replace("No info","No Info")
        
    )    
    .drop(columns="route")
    )

In [34]:
flights_cleaned=clean_data(flights)
flights_cleaned

airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170            0         No Info   3897  
1         13:15:00       445            2         No Info   7662  
2         04:25:00      1140            2         No Info  13882  
3         23:30:00       325            1         No Info   6218  
4         21:35:00       285            1         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150            0         No Info   4107  
10679     23:20:00       155            0         No Info   4145  
10680     11:20:00       180            0         No Info   7229  
10681     14:10:00       160            0         No Info  12648  
10682     19:15:00       500            2         No Info  11753  

[10461 rows x 10 columns]

## Split the Data

In [35]:
flights_final=flights_cleaned
X=flights_final.drop(columns="price")
y=flights_final.price.copy()

In [36]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape,y_train.shape)
print(X_val.shape,y_val.shape)
print(X_test.shape,y_test.shape)

(6694, 9) (6694,)
(1674, 9) (1674,)
(2093, 9) (2093,)


## 7. Export the Subsets

In [37]:
def export_data(X,y,name):
    file_name=f"{name}.csv"
    file_path=os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    
    X.join(y).to_csv(file_path,index=False)
    
    return pd.read_csv(file_path).head()

In [38]:
export_data(X_train,y_train,"train")

airline date_of_journey   source destination  dep_time arrival_time  \
0  Jet Airways      2019-05-27    Delhi      Cochin  20:55:00     12:35:00   
1  Jet Airways      2019-06-12  Kolkata    Banglore  18:55:00     16:20:00   
2    Air India      2019-05-18    Delhi      Cochin  09:45:00     09:25:00   
3       Indigo      2019-06-03   Mumbai   Hyderabad  21:20:00     22:50:00   
4  Jet Airways      2019-04-01   Mumbai   Hyderabad  02:55:00     04:20:00   

   duration  total_stops              additional_info  price  
0       940            1  In-flight meal not included  12898  
1      1285            1                      No Info  13044  
2      1420            2                      No Info  10975  
3        90            0                      No Info   2227  
4        85            0                      No Info   5678

In [39]:
export_data(X_val,y_val,"validation")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-05-27     Delhi      Cochin  09:00:00     19:00:00   
1  Jet Airways      2019-05-24   Kolkata    Banglore  18:55:00     10:05:00   
2  Jet Airways      2019-03-18  Banglore   New Delhi  21:25:00     09:30:00   
3     Spicejet      2019-06-27   Chennai     Kolkata  17:45:00     20:05:00   
4     Air Asia      2019-05-15   Kolkata    Banglore  07:35:00     19:25:00   

   duration  total_stops               additional_info  price  
0       600            1   In-flight meal not included  10675  
1       910            1   In-flight meal not included   8586  
2       725            1                       No Info  13555  
3       140            0  No check-in baggage included   3543  
4       710            1                       No Info   5192

In [40]:
export_data(X_test,y_test,"test")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-03-06  Banglore   New Delhi  08:00:00     08:15:00   
1     Spicejet      2019-06-06   Kolkata    Banglore  22:20:00     00:40:00   
2       Indigo      2019-03-18   Kolkata    Banglore  05:30:00     08:20:00   
3  Jet Airways      2019-03-24    Mumbai   Hyderabad  15:50:00     17:20:00   
4     Spicejet      2019-04-27  Banglore       Delhi  09:30:00     12:20:00   

   duration  total_stops              additional_info  price  
0      1455            1                      No Info  17996  
1       140            0                      No Info   3873  
2       170            0                      No Info   4462  
3        90            0  In-flight meal not included   2228  
4       170            0                      No Info   4991